File: 21-merge-two-sorted-lists/merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* merge(ListNode* list1, ListNode* list2, ListNode* curr)
    {
        if(!list1){
            curr->next = list2;
            curr = curr->next;
            return curr;
        }
        if(!list2){
            curr->next = list1;
            curr = curr->next;
            return curr;
        }
        if(list1->val < list2->val)
        {
            curr->next = list1;
            curr = curr->next;
            curr->next = merge(list1->next, list2, curr);
        }
        else
        {
            curr->next = list2;
            curr = curr->next;
            curr->next = merge(list1, list2->next, curr);
        }
        return curr;
    }
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {

        ListNode* prev = new ListNode();
        ListNode* curr = prev;
        merge(list1, list2, curr);
        return prev->next;
    }
};

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        
        ListNode prev = new ListNode();
        ListNode curr = new ListNode();
        prev.next = curr;
        while(list1!=null || list2!=null)
        {
            int val1 = (list1 == null)? Integer.MAX_VALUE : list1.val;
            int val2 = (list2 == null)? Integer.MAX_VALUE : list2.val;
            if(val1 <= val2)
            {
                curr.next = list1;
                list1 = list1.next;
            }
            else
            {
                curr.next = list2;
                list2 = list2.next;
            }
            curr = curr.next;
        }
        return prev.next.next;
    }
}